src/Compilers/CSharp/Portable/Compiler/TypeCompilationState.cs

```diff
@@ -164,6 +172,59 @@ public void Free()
             }
 
             _wrappers = null;
+            _constructorInitializers = null;
+        }
+
+        // Add an edge to the ctor-initializer graph
+        internal void NoteConstructorInitializer(MethodSymbol method1, MethodSymbol method2, CSharpSyntaxNode syntax)
+        {
+            if (_constructorInitializers == null)
+            {
+                _constructorInitializers = new Dictionary<MethodSymbol, Tuple<MethodSymbol, CSharpSyntaxNode>>();
+            }
+
+            _constructorInitializers.Add(method1, new Tuple<MethodSymbol, CSharpSyntaxNode>(method2, syntax));
+        }
+
+        // Report any cycles in the ctor-intializer graph starting at the given ctor.
+        internal void ReportCtorInitializerCycles(MethodSymbol ctor, DiagnosticBag diagnostics)
+        {
+            if (_constructorInitializers == null)
+            {
+                return;
+            }
+
+            var visiting = PooledHashSet<MethodSymbol>.GetInstance();
+            var todo = PooledHashSet<MethodSymbol>.GetInstance();
+            todo.AddAll(_constructorInitializers.Keys);
+
+            if (HasCycleAt(ctor, visiting, todo))
+            {
+                var errorLocation = _constructorInitializers[ctor].Item2.Location;
+                diagnostics.Add(ErrorCode.ERR_RecursiveConstructorCall2, errorLocation, ctor);
+                _constructorInitializers.Remove(ctor);
+            }
+
+            visiting.Free();
+            todo.Free();
+        }
+
+        private bool HasCycleAt(MethodSymbol ctor, HashSet<MethodSymbol> visiting, HashSet<MethodSymbol> todo)
+        {
+            // Consider rewriting this code to use an explicit stack instead of recursion.
+            try
+            {
+                if (!visiting.Add(ctor))
+                    return true;
+                if (!todo.Remove(ctor))
+                    return false;
+                var next = _constructorInitializers[ctor];
```

Comment:
It was intended to reduce the complexity of the cycle detection algorithm by preventing the analysis of a ctor that has already been analyzed. `todo` is monotonically growing, while `visited` only contains the ctors in the current chain.